src/s3/util/fileViewerManager.ts

```diff
@@ -78,86 +92,135 @@ export class S3FileViewerManager {
-                //set the S3Tab.editor to undefined
-                if (value.editor) {
-                    if (!editorSet.has(value.editor)) {
-                        value.editor = undefined
-                    }
-                }
-            }
-        })
+    public async focusAndCloseTab(
+        uri: vscode.Uri,
+        editor?: vscode.TextEditor,
+        workspace = vscode.workspace
+    ): Promise<void> {
+        if (!editor) {
+            const doc = await workspace.openTextDocument(uri)
+            await this.window.showTextDocument(doc, {
+                preview: false,
+            })
+        } else {
+            await this.window.showTextDocument(editor.document, {
+                preview: false,
+                viewColumn: editor.viewColumn,
+            })
+        }
 
-        vscode.workspace.onDidCloseTextDocument(async closedTextDoc => {
-            if (this.activeTabs.has(closedTextDoc.uri.fsPath)) {
-                const closedTab = this.activeTabs.get(closedTextDoc.uri.fsPath)
-                closedTab!.editor = undefined
-            }
-        })
+        await vscode.commands.executeCommand('workbench.action.closeActiveEditor')
     }
 
     /**
      * Given an S3FileNode, this function:
      * Checks and creates a cache to store downloads
      * Retrieves previously cached files on cache and
      * Downloads file from S3 ands stores in cache
-     * Opens the tab on read-only with the use of an S3Tab
+     * Opens the tab on read-only with the use of an S3Tab, or shifts focus to an edit tab if any.
      *
      * @param fileNode
-     * @returns
      */
-    public async openTab(fileNode: S3FileNode): Promise<void> {
+    public async openInReadMode(fileNode: S3FileNode): Promise<void> {
+        if (fileNode.file.sizeBytes! < SIZE_LIMIT) {
+            this.toPreview = fileNode.file.arn
+        }
+
         getLogger().verbose(`S3FileViewer: Retrieving and displaying file: ${fileNode.file.key}`)
         showOutputMessage(
             localize('AWS.s3.fileViewer.info.fileKey', 'Retrieving and displaying file: {0}', fileNode.file.key),
             this.outputChannel
         )
+        const headResponse = await fileNode.s3.getHeadObject({
+            bucketName: fileNode.bucket.name,
+            key: fileNode.file.key,
+        })
 
         const fileLocation = await this.getFile(fileNode)
         if (!fileLocation) {
             return
         }
         getLogger().verbose(`S3FileViewer: File from s3 or temp to be opened is: ${fileLocation}`)
+        const s3Uri = vscode.Uri.parse('s3:' + fileLocation.fsPath)
+
+        //before opening, ask user how to handle it if it is not text
+        const type = mime.contentType(headResponse.ContentType!)
+        const charset = mime.charset(type as string)
+
+        if (charset != 'UTF-8') {
+            const prompt = "Can't open this file type in read-only mode, do you want to try opening in edit?"
```

Comment:
Specify why? Are there other charsets that are valid for VS Code as well? If so, make sure you check those too.